update-docs : add details for how to enable/disable Policy Audit Mode by endpoint (#19876)
cilium-team/cilium: Documentation/gettingstarted/policy-creation.rst

Policy Audit Mode configures Cilium to allow all traffic while logging all
connections that would otherwise be dropped by policy. Policy Audit Mode may be
configured for the entire daemon using ``--policy-audit-mode=true`` or for
individual Cilium Endpoints. When Policy Audit Mode is enabled, no network
policy is enforced so this setting is **not recommended for production
deployment**. Policy Audit Mode supports auditing network policies implemented
at networks layers 3 and 4. This guide walks through the process of creating
policies using Policy Audit Mode.

.. include:: gsg_requirements.rst
.. include:: gsg_sw_demo.rst

> Enable Policy Audit Mode (entire daemon)

Reviewer: Nit: title case, please (here and below)

Author: updated to title case

> ========================================

Reviewer 2: A few lines down, the `#. Enable Policy Audit Mode ...` creates a numbered list because of the `#.`. You can see this live if you scroll to the bottom of the PR, look in the workflows box and scroll down to the "deploy/netlify" workflow and click the "Details" link, then browse to this page.

Now that this is separate, it makes sense to remove the `#.` before that step, otherwise there's just a list of "1. Enable Policy Audit Mode in the daemon" and no other numbered items after it.

Author: removed the `#. ...` step here.

Policy Audit Mode configures Cilium to allow all traffic while logging all
connections that would otherwise be dropped by policy. Policy Audit Mode may be
configured for the entire daemon using ``--policy-audit-mode=true`` or for
individual Cilium Endpoints. When Policy Audit Mode is enabled, no network
policy is enforced so this setting is **not recommended for production
deployment**. Policy Audit Mode supports auditing network policies implemented
at networks layers 3 and 4. This guide walks through the process of creating
policies using Policy Audit Mode.

.. include:: gsg_requirements.rst
.. include:: gsg_sw_demo.rst

Enable Policy Audit Mode (Entire Daemon)
========================================

To observe policy audit messages for all endpoints managed by this daemonset, modify the Cilium configmap and restart all daemons:

> Enable Policy Audit Mode in the daemon

Reviewer: This is no longer needed? It's what the title right above says.

```suggestion
```

Author: removed

         If you installed Cilium via ``helm install``, then you can use ``helm
         upgrade`` to enable Policy Audit Mode:

>          .. code-block:: shell-session

Reviewer: Same remark as Joe's for the `.. pased-literal`, here and for every block with RST substitutions (`|SOMETHING|`).

Author: reverted this change back to "code-block..."

                --namespace $CILIUM_NAMESPACE \\
                --reuse-values \\
                --set policyAuditMode=true

Enable Policy Audit Mode (specific endpoint)
============================================

Cilium can enable Policy Audit Mode for a specific endpoint. This may be helpful when enabling
Policy Audit Mode for the entire daemon is too broad. Enabling per endpoint will ensure other
endpoints managed by the same daemon are not impacted.

This approach is meant to be temporary.  **Restarts of the Cilium pod will reset the Policy Audit
Mode to match the daemon's configuration.**

Policy Audit Mode is enabled for a given endpoint by modifying the endpoint configuration via
the ``cilium`` tool on the endpoint's Kubernetes node. The steps include:

#. Determine the endpoint id on which Policy Audit Mode will be enabled.
#. Identify the Cilium pod running on the same Kubernetes node corresponding to the endpoint.
#. Using the Cilium pod above, modify the endpoint config by setting ``PolicyAuditMode=Enabled``.

The following shell commands performs these steps:

.. code-block:: shell-session

   $ export PODNAME=deathstar

>    $ export NODENAME=$(kubectl get pod -o json | jq -r --arg pod $PODNAME '.items | .[] | select(.metadata.name==$pod) | .spec.nodeName')
>    $ export ENDPOINT=$(kubectl get cep -o json | jq -r --arg pod $PODNAME '.items | .[] | select(.metadata.name==$pod) | .status.id')

Reviewer: Double quotes around `$PODNAME` would be better, and I think the `jq` commands can be simplified as follows (but I' have NOT double-checked):

```suggestion
   $ export NODENAME=$(kubectl get pod -o json | jq -r --arg pod "$PODNAME" '.items[] | select(.metadata.name==$pod).spec.nodeName')
   $ export ENDPOINT=$(kubectl get cep -o json | jq -r --arg pod "$PODNAME" '.items[] | select(.metadata.name==$pod).status.id')
```

Same for the section on disabling policy audit mode.

Author: quoted and simplified

>    $ export CILIUM_POD=$(kubectl -n $CILIUM_NAMESPACE get pod --all-namespaces --field-selector spec.nodeName=$NODENAME -lk8s-app=cilium -o jsonpath='{.items[*].metadata.name}')
>    $ kubectl -n $CILIUM_NAMESPACE exec $CILIUM_POD -c cilium-agent -- cilium endpoint config $ENDPOINT PolicyAuditMode=Enabled

Reviewer: ```suggestion
   $ export CILIUM_POD=$(kubectl -n $CILIUM_NAMESPACE get pod --all-namespaces --field-selector spec.nodeName="$NODENAME" -lk8s-app=cilium -o jsonpath='{.items[*].metadata.name}')
   $ kubectl -n "$CILIUM_NAMESPACE" exec "$CILIUM_POD" -c cilium-agent -- cilium endpoint config "$ENDPOINT" PolicyAuditMode=Enabled
```

Author: quoted shell variables

         .. parsed-literal::

            $ helm upgrade cilium |CHART_RELEASE| \\
                --namespace $CILIUM_NAMESPACE \\
                --reuse-values \\
                --set policyAuditMode=true

Enable Policy Audit Mode (Specific Endpoint)
============================================

Cilium can enable Policy Audit Mode for a specific endpoint. This may be helpful when enabling
Policy Audit Mode for the entire daemon is too broad. Enabling per endpoint will ensure other
endpoints managed by the same daemon are not impacted.

> This approach is meant to be temporary.  **Restarts of the Cilium pod will reset the Policy Audit

Reviewer: ```suggestion
This approach is meant to be temporary.  **Restarting Cilium pod will reset the Policy Audit
```
would read slightly more natural to me.

Author: updated

Mode to match the daemon's configuration.**

Policy Audit Mode is enabled for a given endpoint by modifying the endpoint configuration via
the ``cilium`` tool on the endpoint's Kubernetes node. The steps include:

#. Determine the endpoint id on which Policy Audit Mode will be enabled.
#. Identify the Cilium pod running on the same Kubernetes node corresponding to the endpoint.
#. Using the Cilium pod above, modify the endpoint config by setting ``PolicyAuditMode=Enabled``.

> The following shell commands performs these steps:

Reviewer: ```suggestion
The following shell commands perform these steps:
```

Author: good catch.  fixed.

inbound traffic towards the deathstar to identify that traffic and determine
whether to extend the network policy to allow that traffic.

Apply a default-deny policy:

.. literalinclude:: ../../examples/minikube/sw_deny_policy.yaml

> CiliumNetworkPolicies match on pod labels using an "endpointSelector" to identify

Reviewer: ```suggestion
CiliumNetworkPolicies match on pod labels using an ``endpointSelector`` to identify
```

Author: updated

the sources and destinations to which the policy applies. The above policy denies
traffic sent to any pods with label (``org=empire``). Due to the Policy Audit Mode
enabled above (either for the entire daemon, or for just the ``deathstar`` endpoint),
the traffic will not actually be denied but will instead trigger policy verdict
notifications.

To apply this policy, run:

> .. code-block:: shell-session

Reviewer 2: This will also need to be a parsed-literal in order for the `\ |SCM_WEB|` link to show up correctly.

Author: reverted "code-block..."

         .. parsed-literal::

            $ helm upgrade cilium |CHART_RELEASE| \\
                --namespace $CILIUM_NAMESPACE \\
                --reuse-values \\
                --set policyAuditMode=false

Disable Policy Audit Mode (Specific Endpoint)
=============================================

These steps are nearly identical to enabling Policy Audit Mode.

.. code-block:: shell-session

   $ export PODNAME=deathstar
   $ export NODENAME=$(kubectl get pod -o jsonpath="{.items[?(@.metadata.name=='$PODNAME')].spec.nodeName}")
   $ export ENDPOINT=$(kubectl get cep -o jsonpath="{.items[?(@.metadata.name=='$PODNAME')].status.id}")
   $ export CILIUM_POD=$(kubectl -n "$CILIUM_NAMESPACE" get pod --all-namespaces --field-selector spec.nodeName="$NODENAME" -lk8s-app=cilium -o jsonpath='{.items[*].metadata.name}')
   $ kubectl -n "$CILIUM_NAMESPACE" exec "$CILIUM_POD" -c cilium-agent -- cilium endpoint config "$ENDPOINT" PolicyAuditMode=Disabled
   Endpoint 232 configuration updated successfully

Alternatively, **restarting the Cilium pod** will set the endpoint Policy Audit Mode to the daemon set configuration.

> Verify Policy Audit Mode Disabled
> =================================

Reviewer: ```suggestion
Verify Policy Audit Mode is Disabled
====================================
```

Author: updated....using "title case" here too.
